### PR TITLE
feat: add forgot password feature

### PR DESCRIPTION
**Pull request by Jidekas**
**Description of PR**

This PR:
   - adds forgot password feature
   - updates user model with isVerified, resetToken and resetTokenExpiry fields
   - creates a re-usable email function

---

Please check all that apply.

#### Checklist:

- [x] Rebased with master
- [x] PR Description provided
- [x] Did you communicate these changes with the affected teams?
- [ ] Did unit tests/integration tests/E2E tests get properly added/updated?
- [ ] Is this change production-ready including testing, feature toggles, configurations, and deployment considerations?

#### Frontend Checklist:

- [ ] Is this responsive and mobile friendly?
- [ ] Did you follow the [ARIA specs](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA)

#### Backend Checklist:

- [x] Does the endpoint return the correct success/error format?

---
Screenshot 2023-12-20 at 07 49 38
Screenshot 2023-12-20 at 07 50 05

**Screenshots**  
_add relevant screenshots here_

---

**Test Coverage**  
_add relevant screenshots here_

### CODE REVIEW

**Comment by KristofaJosh** [comment]
Task linked: [CU-86bwj0akd BE - Auth - Forgot Password](https://app.clickup.com/t/86bwj0akd)

File: apps/server/src/controllers/userController.ts [hunk 1]
```diff
@@ -104,3 +107,53 @@ export const registerUser = async (req: Request, res: Response) => {
...
+    })
+    const validate = mailSchema.strict().safeParse(req.body)
+
+    if (validate.success) {
+      let { email } = req.body
+      email = email.trim().toLowerCase()
+      const user = await Users.findOne({ where: { email: email } })
+      const passwordResetData = GenerateOTP()
+      const passwordResetToken = passwordResetData.otp.toString()
+      const tokenExpiery = passwordResetData.expiry
```
Comment (polymathmode) [inline apps/server/src/controllers/userController.ts:124]: You may want to correctly spell the variable in line 124(tokenExpiry)

File: apps/server/.env.sample [hunk 2]
```diff
@@ -5,6 +5,7 @@ DB_PASSWORD=
 DB_USERNAME=
 
 REDIS=//localhost:6379
+APP_HOST=http://localhost
```
Comment (KristofaJosh) [inline apps/server/.env.sample:8]: No need for this....
I have updated the code

File: apps/server/src/config/env.ts [hunk 3]
```diff
@@ -26,8 +26,11 @@ const {
   GMAIL_PASSWORD,
   APP_SECRET,
   REDIS,
+  APP_HOST,
 } = process.env
 
+const baseUrl = `${APP_HOST}:${PORT}`
```
Comment (KristofaJosh) [inline apps/server/src/config/env.ts:32]: No need for this....
I have updated the code

File: apps/server/src/controllers/userController.ts (hunk 1 above)
Comment (KristofaJosh) [reply apps/server/src/controllers/userController.ts:124]: I have updated the code

File: apps/server/src/controllers/userController.ts [hunk 4]
```diff
@@ -104,3 +107,53 @@ export const registerUser = async (req: Request, res: Response) => {
...
+        await Users.update(
+          { resetToken: passwordResetToken, resetTokenExpiry: tokenExpiery },
+          { where: { id: user.id } }
+        )
+        const link = `${ENV.baseUrl}/api/v1/reset-password?userId=${user.id}&token=${passwordResetToken}`
+        sendEmail({
+          firstName: user.firstName,
+          email: email,
+          message: link,
+        })
```
Comment (KristofaJosh) [inline apps/server/src/controllers/userController.ts:136]: should be background job

Comment (KristofaJosh) [reply apps/server/src/controllers/userController.ts:136]: I have updated the code

File: apps/server/src/controllers/userController.ts [hunk 5]
```diff
@@ -104,3 +107,53 @@ export const registerUser = async (req: Request, res: Response) => {
...
+        return res.status(HTTP_STATUS_CODE.SUCCESS).json({
+          message: `password reset link has been sent to your email`,
+        })
+      } else {
+        return res.status(HTTP_STATUS_CODE.NOT_FOUND).send({
+          message: 'Account not found',
+        })
```
Comment (KristofaJosh) [inline apps/server/src/controllers/userController.ts:143]: never let them know we don't have the account, always send 200 regardless

Comment (KristofaJosh) [reply apps/server/src/controllers/userController.ts:143]: I have updated the code

File: apps/server/src/models/users.ts [hunk 6]
```diff
@@ -74,6 +77,19 @@ Users.init(
     otp_expiry: {
       type: DataTypes.DATE,
     },
+    resetToken: {
+      type: DataTypes.STRING,
+      allowNull: true,
+    },
+    resetTokenExpiry: {
```
Comment (KristofaJosh) [inline apps/server/src/models/users.ts:84]: what's otp_expiry ?

File: apps/server/src/utils/mailFunction/mailfunction.ts [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+      html,
+    }
+
+    const info = await transporter.sendMail(mailOptions)
+    console.log('Email sent:' + info.response)
+  } catch (error) {
+    return console.log({ message: 'error sending mail', Error: error })
+    // Handle email sending failure
+  }
+}
```
Comment (KristofaJosh) [inline apps/server/src/utils/mailFunction/mailfunction.ts:56]: moved to bg.jobs

Comment (KristofaJosh) [reply apps/server/src/utils/mailFunction/mailfunction.ts:56]: I have updated the code

**Review by KristofaJosh — CHANGES_REQUESTED** [review]
I have updated the code

File: apps/server/src/models/users.ts (hunk 6 above)
Comment (Jidekas) [reply apps/server/src/models/users.ts]: Thats expiry is  for the email verification otp

**Review by KristofaJosh — APPROVED** [review]

**Pull request merged**